libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+package protocols
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+)
+
+type Address string
+type Performative []string
+type SomeMessageType string
+
+const (
+	NONCE_BYTES_NB = 32
+)
+
+type Role string
+
+const (
+	Role1             Role = "role1"
+	Role2             Role = "role2"
+	StartingMessageId      = 1
+	StartingTarget         = 0
+)
+
+type RuleType struct {
+	initialPerformatives  Performative
+	terminalPerformatives Performative
+	validReplies map[Performative][]Performative
+}
+
+type MessageId int
+type Target int
+type DialogueInterface interface {
+	getSelfAddress() Address
+	getRole() Role
+	getRules() RuleType
+	getMessageClass() AbstractMessage
+	isSelfInitiated() bool
+	getLastIncomingMessage() AbstractMessage
+	getLastOutgoingMessage() AbstractMessage
+	getLastMessage() AbstractMessage
+	counterPartyForMessage(AbstractMessage) bool
+	isMessageBySelf(AbstractMessage) bool
+	isMessageByOther(AbstractMessage) bool
+	getMessage(MessageId) AbstractMessage
+	hasMessageId(MessageId) bool
+
+	update(AbstractMessage)
+	validateNextMessage(AbstractMessage) (bool, string)
+	basicValidations(AbstractMessage) (bool, string)
+	basicValidationInitialMessage(AbstractMessage) (bool, string)
+	basicValidationNonInitialMessage(AbstractMessage) (bool, string)
+	isEmpty() bool
+	updateIncomingAndOutgoingMessages(AbstractMessage)
+	isBelongingToADialogue(AbstractMessage) bool
+
+	validateMessageTarget(AbstractMessage) string
+	validateMessageId(AbstractMessage) string
+	getMessageById(MessageId) AbstractMessage
```

Comment:
To be more faithful with the Python implementation, this method should return a reference to the message:
```suggestion
	getMessageById(MessageId) *AbstractMessage
```

however, I would also accept returning by value (which implies a copy of the returned value)